### PR TITLE
PM-1498 VPos polling API changes

### PR DESCRIPTION
**Pull request by alessandrodossi**
The code pushed with this PR is meant to meet the requirements for the new VPos polling API.
New parameters have been added in response according to swagger specification attached to task PM-1374.
**This PR only contains the work for VPos logic**, thus XPay logic is still in development and will be pushed asap
#### List of Changes

- GET logic implemented
- JUnit tests adapted
#### How Has This Been Tested?
Tests in local environment have been done in order to check the changes

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.  --> planning to do it asap
- [ ] I have created or updated an API interface.
- [ ] I have updated the OpenAPI spec --> planning to do it asap
- [ ] I have created or updated a database table
- [ ] I have updated both the Oracle and Postgres migrations
- [x] My PR is cohesive and self-contained (rule of thumb: less than 250 changed lines) 

### Link to story

https://pagopa.atlassian.net/browse/PM-1498

### CODE REVIEW

File: src/main/java/it/pagopa/pm/gateway/dto/enums/ThreeDS2ResponseTypeEnum.java [hunk 1]
```diff
@@ -1,7 +1,18 @@
...
     METHOD,
     CHALLENGE,
-    AUTHORIZATION
+    AUTHORIZATION,
+    ERROR;
+
+    public static ThreeDS2ResponseTypeEnum getEnumFromValue(String value) {
+        return Arrays.stream(ThreeDS2ResponseTypeEnum.values())
+                .filter(enumValue -> StringUtils.equals(enumValue.name(), value))
+                .findFirst().orElse(null);
```
[the remark this answers is not in the document]
Comment (alessandrodossi) [reply src/main/java/it/pagopa/pm/gateway/dto/enums/ThreeDS2ResponseTypeEnum.java]: Not really, you're right. I pushed the change

File: src/main/java/it/pagopa/pm/gateway/dto/enums/ThreeDS2ResponseTypeEnum.java [hunk 2]
```diff
@@ -1,7 +1,18 @@
...
     CHALLENGE,
-    AUTHORIZATION
+    AUTHORIZATION,
+    ERROR;
+
+    public static ThreeDS2ResponseTypeEnum getEnumFromValue(String value) {
+        return Arrays.stream(ThreeDS2ResponseTypeEnum.values())
+                .filter(enumValue -> StringUtils.equals(enumValue.name(), value))
+                .findFirst().orElseThrow(() -> new IllegalArgumentException("Invalid response type: " + value));
+    }
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/dto/enums/ThreeDS2ResponseTypeEnum.java:17]: Can be deleted, identical to `ThreeDsResponseTypeEnum.valueOf(String)`

File: src/main/java/it/pagopa/pm/gateway/service/CcPaymentInfoService.java [hunk 3]
```diff
@@ -40,46 +50,74 @@ public CcPaymentInfoService(@Value("${vpos.method.notifyUrl}") String methodNoti
...
-            } else if (CHALLENGE.name().equals(paymentInfo.getResponseType())) {
-                String creq = getCreqFromChallengeUrl(paymentInfo);
+        String authorizationUrl = paymentRequestEntity.getAuthorizationUrl();
+        String[] vposUrlAndCreq = StringUtils.splitByWholeSeparatorPreserveAllTokens(authorizationUrl, CREQ);
+        String vposUrl = null;
+        String creq = null;
+        if (ObjectUtils.isNotEmpty(vposUrlAndCreq)) {
+            vposUrl = vposUrlAndCreq[0];
+            creq = vposUrlAndCreq.length == 2 ? vposUrlAndCreq[1] : null;
+        }
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/CcPaymentInfoService.java:78]: This snippet manipulates a URL via its string representation and makes the underlying logic a bit less clear (for example here we are assuming that `creq` is optional and can be the only query parameter in the authorization url).
I would suggest to either refactor this to extract the `creq` query param via URL/URI parsing (e.g. see [here](https://stackoverflow.com/questions/13592236/parse-a-uri-string-into-name-value-collection)) or extract this to a separate private method

File: src/main/java/it/pagopa/pm/gateway/service/VposService.java [hunk 4]
```diff
@@ -65,7 +65,7 @@ public StepZeroResponse startCreditCardPayment(String clientId, String mdcFields
         }
 
         String idTransaction = request.getIdTransaction();
-        log.info(String.format("START - POST %s for idTransaction %s", REQUEST_PAYMENTS_VPOS, idTransaction));
+        log.info(String.format("START - POST %s for idTransaction %s", VPOS_AUTHORIZATIONS, idTransaction));
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/VposService.java:68]:
```suggestion
        log.info(String.format("START - POST {} for idTransaction {}", VPOS_AUTHORIZATIONS, idTransaction));
```

File: src/main/java/it/pagopa/pm/gateway/service/VposService.java [hunk 5]
```diff
@@ -102,7 +102,7 @@ private StepZeroResponse createStepZeroResponse(String errorMessage, String requ
             response.setError(errorMessage);
         }
 
-        log.info(String.format("END - POST %s for requestId %s", REQUEST_PAYMENTS_VPOS, requestId));
+        log.info(String.format("END - POST %s for requestId %s", VPOS_AUTHORIZATIONS, requestId));
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/VposService.java:105]:
```suggestion
        log.info(String.format("END - POST {} for requestId {}", VPOS_AUTHORIZATIONS, requestId));
```

File: src/main/java/it/pagopa/pm/gateway/service/CcPaymentInfoService.java [hunk 6]
```diff
@@ -40,46 +53,96 @@ public CcPaymentInfoService(@Value("${vpos.method.notifyUrl}") String methodNoti
...
+        return uriBuilder.build().toString();
+    }
+
+    private String getCreqFromAuthUrl(String authorizationUrl) throws URISyntaxException {
+        URIBuilder uriBuilder = new URIBuilder(authorizationUrl);
+        NameValuePair creqPair = uriBuilder.getQueryParams().stream()
+                .filter(queryParam -> StringUtils.equals(queryParam.getName(), CREQ))
+                .findFirst()
+                .orElse(null);
+        return Objects.nonNull(creqPair) ? creqPair.getValue() : null;
```
Comment (giovanniberti) [inline src/main/java/it/pagopa/pm/gateway/service/CcPaymentInfoService.java:119]:
```suggestion
        return uriBuilder.getQueryParams().stream()
                .filter(queryParam -> StringUtils.equals(queryParam.getName(), CREQ))
                .findFirst()
                .map(NameValuePair::getValue)
                .orElse(null);
```

**Review by giovanniberti — APPROVED** [review]

**Pull request merged**